### PR TITLE
lib: mempool: Fix broken algorithm

### PR DESCRIPTION
**Pull request by pdunaj**
This change should make allocator usable until mempool algorithms will
be properly fixed.

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by pdunaj** [comment]
Build and tested on PCA20041. OOM issue not replicated anymore after patch is applied.

**Comment by carlescufi** [comment]
@nashif this is a bug in the mempool module of the kernel

**Comment by nashif** [comment]
> This change should make allocator usable until mempool algorithms will
> be properly fixed.

can you provide more details, is there a bug open about that?

**Comment by pdunaj** [comment]
> > This change should make allocator usable until mempool algorithms will
> > be properly fixed.
> 
> can you provide more details, is there a bug open about that?

#11022

**Review by tyrantlizard77 — DISMISSED** [review]
LGTM, would like to get @andyross opinion.

**Comment by andrewboie** [comment]
We seem to be getting some fatal exceptions in Nios II and Xtensa for the mem_pool tests though

**Comment by andyross** [comment]
Note that this looks like it's going to increase interrupt-locked latencies significantly, which was the point of the admittedly somewhat scary-looking synchronization to begin with.

I'm not really understanding the bug being fixed, though.   The linked bug has two cases: one is a spurious -ENOMEM where two contended threads are hitting the allocator.  We'll be "out of memory" instantaneously due to internal fragmentation even though waiting a bit would see that memory available again.  But... so what?  That behavior is true in some form of all allocators!  We really were "out of memory" at the instant the second thread tried to find some.  Let the higher layers handle retries.

The second case seems to be implying some interaction with thread priority like live lock or something?  Honestly I don't understand it.

Basically: what's failing that this fixes, and how do we see that in code?  The cost is nontrivial.

**Comment by pdunaj** [comment]
Hi @andyross , I am not the author of the mempool but my understanding is that it was assumed that when there is a collision (i.e. one context will grab the huge block or 4 mid blocks), context calling malloc will get the `-EAGAIN` and then, on the next try, will get the memory it wants.

There are following problems here:
1) if block (during split) or blocks (during merge) form the whole pool size, `-ENOMEM` will be returned and retry will not happen. This could be fixed by tracking the real usage of the memory and by returning `-EAGAIN` in collision case while returning `-ENOMEM` in true nomem. But there is a bigger problem here.
2) There is absolutely no guarantee that you will get the memory from the system when it is available even if you loop (i.e. algorithm is unstable). But there is even a bigger problem when you will not get the memory even if it is available.
3) When the context calling alloc has higher priority then the one calling free/malloc while collision occurs, first one can loop however many times it wants. The lower priority context will never get the CPU to finish block merging/splitting and memory will not be returned. Note that there is plenty of memory available in the system (entire pool to be precise) but algorithm hides it for no good reason.

Why this is a problem? Even if you use small portion of the memory (i.e. smallest possible chunk) this algorithm is causing split on every alloc and merge on every free. This means that every time memory being freed entire pool is taken out. E.g. if I have 4kB heap and alloc 32B this code will cause 4kB to disappear for a moment. E.g. consider following: An (alloc from thread n), Fn (free from thread n).
```
A0* A1 F1 F2* A0* A1 F1 F2* 
```
(* - for a moment entire heap memory disappears)
Imaging  now that you allocate a small chunk from an interrupt at position with *. There is a chance that memory won't be there. You can increase heap size as much as you like - memory won't be there as it is locked out for the time when split/merge is performed. As this is done by low priority thread this won't be finished as long as you execute.

This algorithm will work only if something grabs the memory for a long time and so will prevent split/merge from happening. This is however a rather unexpected case. I think that in embedded world if you know memory is needed for entire lifespan of the program you reserve it statically. Pools like that are useful mostly when you create short lived objects (often in bursts).

We use pool for creating events. When user starts interacting with the device events are created, handled and freed. Pool is empty for most of the time (when device is idle). We have several points of interaction (i.e. user can press, move, rotate several things on the device). Some of this points are readout by thread and some are obtained by interrupt - events are created from thread and irq context. When user causes random, rapid interaction on several points you get OOM even though entire memory is free. You don't have to wait long - few seconds are enough.

I expect that creating code for replication should be easy.

**Review by andyross — CHANGES_REQUESTED** [review]
You're asking too much here.  One of the primary design goals of mempool, as distinct from general heaps like the malloc in newlib, is that it works cleanly in uninterruptable contexts like ISRs.  This requires careful latency control, which gets pretty badly defeated by this patch.

Ultimately what's happening here is that you have two threads contending on heap state, and indeed that can temporarily leave the heap in a situation where there is "enough" memory in an ideal sense, but in a practical sense it's not available.  This is a property of all heaps!  Mempool isn't alone here.  You can fragment your way to a NULL malloc return in a Linux process with gigabytes of memory too.

If you want a 100% reliable allocation, Zephyr already gives you two options:

1. You can use a blocking allocation call, which is guaranteed to give up the CPU for lower priority threads, fixing the priority inversion issue above.  Obviously that involves a timeout in -ENOMEM situations, though with a little work that can probably be fixed too.

2. You can just use newlib's malloc(), which is a traditional heap with traditional serialization properties (and comcommitantly MUCH higher locking latencies) that probably match your expectations better.

I think I'm going to -1 this absent a really clear bug exerciser that can't be fixed in other ways, or if you can show that the worst case interrupt latency isn't as badly affected as you think.

**Comment by pdunaj** [comment]
Hi @andyross , I give up. I already went this path and see no sense in persuading people that have their minds fixed. If Zephyr's community wants to have a broken allocator then feel free to keep it.

Knowing what is the problem I can work around it in the application.
My only advice for you is to extensively document that `k_malloc` can randomly return NULL when entire heap is free. I doubt anybody will use it for critical systems but millions of randomly crashing devices would be bad enough.

**Comment by andyross** [comment]
OK, @pizi-nordic sat me down via a video chat and explained the circumstance I'm missing: an allocation out of an ISR (totally within the design intent) can have interrupted an existing allocator that had removed visible blocks from the free list and see a spurious ENOMEM in a context where you (obviously) can't block or yield to resolve the contention and where you can't use newlib.  That needs to be fixed for sure.

Still -1 here because of the latency concerns.  I think I can get a proper fix in today if I'm lucky.

**Comment by pdunaj** [comment]
@andyross thanks, this is good news :)

As for -1 it is fine. This PR contains a temporary fix. If proper rework will be done right away then I can close it.

Side note on latency. I think it would be good to not merge back the blocks on free but rather try gathering them when allocation of larger piece is requested. More common case seems to be a frequent allocation of small chunks of memory. In such case splitting and merging occurs very often and cause performance penalty.

**Review by pfalcon — CHANGES_REQUESTED** [review]
From the commit message, it's unclear which problems there're with mempool algorithms. Should either add more info or reference a ticket with such info.
